Pd/metric grouping

pdunn opened the pull request:
Provide facilities for controlling the granularity of metrics reporting within an instrumented class and across classes that share a MetricsRegistry.

@schlosna , can you take look here?  This differs from tagging in that tagging allows you to differentiate reporters of a shared metric name, while this allows for the bundling of instrumented calls into a shared metric.  The annotation facilities could be leveraged to provide additional handling for Method (static) and Parameter (dynamic) tags.

Commit:
Initial add of metric group annotation

Commit:
Add class level default to handling

Commit:
Utility to scan class for annotations on parent classes and interfaces

Commit:
Utility to scan class for annotations on parent classes and interfaces

Commit:
Utility to scan class for annotations on parent classes and interfaces

Commit:
Fix parent annotation tracking inside Handler and tests

Commit:
test variations for AnnotationHelper

Commit:
document annotation helper

Commit:
Implement global prefix across instrumented interfaces

Commit:
[DO NOT MERGE] Incomplete final wiring of MetricGroup handling

Commit:
Better method name resolution

Commit:
Merge remote-tracking branch 'remotes/origin/develop' into pd/metric-grouping

Commit:
checkstyle fixes

Commit:
standardize on use of MethodSignature

schlosna commented on tritium-lib/src/test/java/com/palantir/tritium/TritiumTest.java:
```diff
@@ -73,6 +73,10 @@ public void testInstrument() {
     }
 
+    public void testInstrumentWithPrefix() {
+
```
- [ ] TODO?

schlosna commented on tritium-lib/src/main/java/com/palantir/tritium/proxy/Instrumentation.java:
```diff
@@ -118,14 +118,20 @@ private Builder(Class<T> interfaceClass, U delegate) {
         }
 
-        public Builder<T, U> withMetrics(MetricRegistry metricRegistry) {
+        public Builder<T, U> withMetrics(MetricRegistry metricRegistry, String globalPrefix) {
```
We should add JavaDoc for what the global prefix entails

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/MetricsInvocationEventHandler.java:
```diff
@@ -42,16 +47,55 @@
 
+    //consider creating annotation handlers as separate objects
+    private final Map<AnnotationHelper.MethodSignature, String> metricGroups;
+    private final String globalGroupPrefix;
```
nit: mark `@Nullable private final String globalGroupPrefix;`

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/MetricsInvocationEventHandler.java:
```diff
@@ -42,16 +47,55 @@
     private final MetricRegistry metricRegistry;
     private final String serviceName;
 
+    //consider creating annotation handlers as separate objects
```
+1, I think we'll want to have mapping of method signature to handler to implement https://github.com/palantir/tritium/issues/14

pdunn replied:
I'm thinking about ways to do this.  Most clean way seems to be:

1) MetricsInvocationEventHandler has a Multimap<MethodSignature, MetricsAnnotationHandler>
2) Each MetricsAnnotationHandler contains a Map<MethodSignature, Annotation>

Sort of begs the question if the MetricsInvocationEventHandler needs awareness of a the Annotation Handlers or whether the annotation handlers can be completely generic invocation handlers.

schlosna replied:
yeah, I think we can leave this as-is for now and extract flesh out the annotation handlers separately

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/annotations/AnnotationHelper.java:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        return builder.addAll(Arrays.asList(clazz.getInterfaces())).build();
+    }
```
Wondering if we need to traverse the full super-type hierarchy with something like:

```
    @VisibleForTesting
    static Set<Class<?>> getAllParentClasses(Class<?> clazz) {
        ImmutableSet.Builder<Class<?>> builder = ImmutableSet.builder();
        builder.add(clazz.getInterfaces());
        Class<?> superclass = clazz.getSuperclass();
        while (superclass != null) {
            builder.add(superclass.getInterfaces());
            builder.add(superclass);
            superclass = superclass.getSuperclass();
        }
        return builder.build();
    }
```

pdunn replied:
Both times this used it's accessed in recursive methods that climb the type hierarchy.  I'm not sure it's needed and the behavior on on which interface is selected for annotations becomes the parent overwriting the child.

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/annotations/AnnotationHelper.java:
```diff
@@ -0,0 +1,156 @@
+
+        private MethodSignature(String methodName, Class<?>... parameterTypes) {
+            this.methodName = checkNotNull(methodName);
+            this.parameterTypes = parameterTypes != null ? parameterTypes : new Class<?>[]{};
```
nit: we probably need to clone the parameter types

```
private static final Class<?>[] NO_ARGS = new Class<?>[0];

this.parameterTypes = (parameterTypes == null || parameterTypes.length == 0) ? NO_ARGS : parameterTypes.clone();
```

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/MetricsInvocationEventHandler.java:
```diff
@@ -42,16 +47,55 @@
+
+        for (Method method : serviceClass.getMethods()) {
+            AnnotationHelper.MethodSignature sig = AnnotationHelper.MethodSignature.of(method);
+            MetricGroup methodGroup = AnnotationHelper.getMethodAnnotation(MetricGroup.class, serviceClass, sig);
```
see other comment

```
            MetricGroup methodGroup = AnnotationHelper.getAnnotationIfPresent(MetricGroup.class, serviceClass, method);
```

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/annotations/AnnotationHelper.java:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        return null;
+    }
```
```
    /**
     * Depth first search up the Type hierarchy to find a matching annotation,  Types which do not implement the
     * specified method signature are ignored.
     *
     * @param annotation annotation class
     * @param clazz containing class
     * @param method method
     * @param <T> annotation type
     * @return - First found matching annotation or null
     */
    @Nullable
    public static <T extends Annotation> T getAnnotationIfPresent(Class<T> annotation, Class<?> clazz, Method method) {
        return getAnnotationIfPresent(annotation, clazz, method, method.getParameterTypes());
    }

    /**
     * Depth first search up the Type hierarchy to find a matching annotation. Types which do not implement the
     * specified method signature are ignored.
     *
     * @param annotationClass annotation class
     * @param clazz containing class
     * @param method method
     * @param parameterTypes method parameter types
     * @param <T> annotation type
     * @return - First found matching annotation or null
     */
    @Nullable
    private static <T extends Annotation> T getAnnotationIfPresent(
            Class<T> annotationClass,
            Class<?> clazz,
            Method method,
            Class... parameterTypes) {

        if (method == null) {
            return null;
        }

        T maybeAnnotation = method.getAnnotation(annotationClass);
        if (maybeAnnotation != null) {
            return maybeAnnotation;
        }

        String methodName = method.getName();
        for (Class<?> parentClass : getAllParentClasses(clazz)) {
            for (Method parentMethod : parentClass.getMethods()) {
                if (methodName.equals(parentMethod.getName())) {
                    Class<?>[] parentMethodParamTypes = parentMethod.getParameterTypes();
                    if (Arrays.equals(parameterTypes, parentMethodParamTypes)) {
                        maybeAnnotation = parentMethod.getAnnotation(annotationClass);
                        if (maybeAnnotation != null) {
                            log.info("Found {} on matching method {} on {}",
                                    maybeAnnotation, parentMethod, parentClass);
                            return maybeAnnotation;
                        }
                    }
                }
            }
        }

        return null;
    }
```

pdunn replied:
@schlosna , are you advocating ditching the MethodSignature as the argument here and going back to either (Method) or (String, Class[])?

schlosna replied:
I had branched off your previous changes and was testing out some things. We probably need some test cases where method name is the same but signatures differ by argument types.

schlosna commented on tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/annotations/AnnotationHelper.java:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        public Class<?>[] getParameterTypes() {
+            return parameterTypes;
```
I think we can get away without exposing the getters, and probably move this to a top level class.

```
public final class MethodSignature {

    private final String name;
    private final List<Class<?>> parameterTypes;

    private MethodSignature(String name, List<Class<?>> parameterTypes) {
        this.name = name;
        this.parameterTypes = ImmutableList.copyOf(parameterTypes);
    }

    public static MethodSignature from(Method method) {
        return new MethodSignature(method.getName(), ImmutableList.copyOf(method.getParameterTypes()));
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof MethodSignature)) {
            return false;
        }

        MethodSignature that = (MethodSignature) obj;
        return name.equals(that.name) && parameterTypes.equals(that.parameterTypes);
    }

    @Override
    public int hashCode() {
        return 31 * name.hashCode() + parameterTypes.hashCode();
    }

    @Override
    public String toString() {
        return "MethodSignature{"
                + "methodName='" + name + '\''
                + ", parameterTypes=" + parameterTypes
                + '}';
    }

}
```

schlosna commented on tritium-metrics/src/test/java/com/palantir/tritium/event/metrics/MetricsInvocationEventHandlerTest.java:
```diff
@@ -69,4 +93,54 @@ public void testSystemPropertySupplier_Handler_Enabled() throws Exception {
+    private void callVoidMethod(
+            MetricsInvocationEventHandler handler, Object obj, String methodName, boolean success) throws Exception {
+
+        Method method = obj.getClass().getMethod(methodName);
```
nit: do you want to use this in line below?

schlosna commented on tritium-lib/src/main/java/com/palantir/tritium/Tritium.java:
```diff
@@ -44,4 +44,22 @@ private Tritium() {
+     * @return instrumented proxy implementing specified service interface
+     */
+    public static <T, U extends T> T instrument(
+            Class<T> serviceInterface, U delegate, MetricRegistry metricRegistry, String globalPrefix) {
```
I'm leaning toward not adding this method here in `Tritium` as it is probably clearer for folks that want control over global prefix to use the `Instrumentation` builder explicitly

```
Instrumentation.builder(serviceInterface, delegate)
                .withMetrics(metricRegistry, globalPrefix)
                .withPerformanceTraceLogging()
                .build();
```

pdunn replied:
cool, can remove that

Commit:
First pass on code review

Document global prefix
Remove unused test method call
Add Nullable to globalPrefix field
Better hierarchy traversal for parent class types
remove global prefix from the Tritium static

Commit:
checkstyle and additional unit tests

Commit:
checkstyle and code review
